pandas/tests/tseries/test_timezones.py

```diff
@@ -1229,6 +1229,26 @@ def f():
         dt = Timestamp('2013-11-03 01:59:59.999999-0400', tz='US/Eastern')
         assert dt.tz_localize(None) == dt.replace(tzinfo=None)
 
+    def test_replace_across_dst(self):
```

Comment:
can you tests with dateutil as well

Reply:
The test is pretty specific to pytz API

Reply:
The test is specific to the `pytz` API because the *problem* is specific to the `pytz` API, so you can't write tests that currently fail with `dateutil`, *but* you can, in fact, write this test to be agnostic as to whether the test takes a `pytz` or `dateutil` zone.

Rather than using `localize` and `normalize`, express the "expected" datetimes as UTC datetimes and use `.astimezone`. `pytz` and `dateutil` support `astimezone` (and, in fact, `pytz.normalize` essentially just uses `.astimezone` under the hood). Since `pandas` provides their own API layer on top of `datetime`, it is *also* agnostic to the timezone provider.

Reply:
> Rather than using localize and normalize, express the "expected" datetimes as UTC datetimes and use .astimezone

Is this a suggestion for this test or more generally?

Reply:
> Is this a suggestion for this test or more generally?

In this test at least, since with tests you can always engineer your datetime literals such that you never have to use anything except `astimezone`.

It's also just a useful piece of information to have, that `astimezone` is one of the few timezone functions where `pytz` behaves more or less nicely. You only need `normalize` semantics with `pytz` for things like `replace` or "calendar" arithmetic, where you want to change the naive portion of the datetime to a specific value and then change the time zone offset to match. If you want "absolute" arithmetic (where you want to go forward a certain number of hours or seconds or something), then `the_operation(dt.astimezone(UTC)).astimezone(dt.tzinfo)` will always give the right answer.

In this case you know what the correct answer *should* be, in absolute time, so you can just declare your initial variable as the correct answer in `UTC` and convert that to the time zone you care about. If you do it generically like that, you are insulated from any weird quirks of `pytz`'s interface, and you get `dateutil` support for free (so you can parametrize this test using `pytz` and `dateutil` zones).

Of course, this method will fail if you try to construct an imaginary time, since there is no mapping between UTC and imaginary times.